tiledb/sm/misc/resource_pool.h

```diff
@@ -139,6 +144,92 @@ class ResourcePool {
   friend class ResourceHandle<T, tiledb::sm::ResourcePool>;
 };
 
+template <class T>
+class BlockingResourcePool {
+ public:
+  using resource_handle = ResourceHandle<T, tiledb::sm::BlockingResourcePool>;
+
+  /** Constructor. */
+  BlockingResourcePool(unsigned int n)
+      : resources_(n)
+      , unused_(n)
+      , unused_idx_(n - 1)
+      , num_blocked_threads_(0) {
+    for (unsigned int i = 0; i < n; i++)
+      unused_[i] = i;
+  }
+
+  /** Take a resource from the pool. */
+  resource_handle take() {
+    if (resources_exhausted()) {
+      // block until available again
+      std::unique_lock<std::mutex> e_lck(exhaustion_mtx_);
+      num_blocked_threads_++;
+      exhaustion_cv_.wait(e_lck, [this]() { return !resources_exhausted(); });
```

Comment:
@eric-hughes-tiledb should we use `wait_for` instead so that we can set a timeout and then act upon it (e.g. throw?) to avoid a thread stuck for ever for some reason?

Reply:
FWIW, be aware that even apart from misbehavior it's possible to have a 'stuck' thread in a heavily loaded system, that environments do not necessarily guarantee a waking/servicing order [first-sleep first-wake might often be desireable but is not guaranteed], and so if a process is in a load situation where you repeatedly have, for example, at least two threads in need of a resource, there's a possibility that one will always be served and the other left starving.  Might not even take a heavily loaded system, might just depend on the characteristics of the particular process...